File: lib/xero_cli/oauth.ex
defmodule XeroCLI.OAuth do
  @moduledoc """
  Handles OAuth 2.0 authentication flow for Xero API.
  Uses PKCE (Proof Key for Code Exchange) for security.
  """

  @xero_auth_url "https://login.xero.com/identity/connect/authorize"
  @xero_token_url "https://identity.xero.com/connect/token"
  @xero_connections_url "https://api.xero.com/connections"
  @redirect_uri "http://localhost:8888/callback"

  def get_auth_url(client_id) do
    # Generate PKCE code verifier and challenge
    code_verifier = generate_code_verifier()
    code_challenge = generate_code_challenge(code_verifier)

    # Store code verifier for later use
    XeroCLI.Config.set_setting("pkce_verifier", code_verifier)

    params = %{
      "response_type" => "code",
      "client_id" => client_id,
      "redirect_uri" => @redirect_uri,
      "scope" =>
        "offline_access openid profile email accounting.transactions accounting.contacts accounting.settings",
      "code_challenge" => code_challenge,
      "code_challenge_method" => "S256"
    }

    query_string = URI.encode_query(params)
    "#{@xero_auth_url}?#{query_string}"
  end

  def exchange_code(code, client_id) do
    # Retrieve stored code verifier
    code_verifier = XeroCLI.Config.get_setting("pkce_verifier")

    if code_verifier == nil do
      {:error, "Code verifier not found. Please restart the authentication process."}
    else
      params = %{
        "grant_type" => "authorization_code",
        "client_id" => client_id,
        "code" => code,
        "redirect_uri" => @redirect_uri,
        "code_verifier" => code_verifier
      }

      body = URI.encode_query(params)

      headers = [
        {"content-type", "application/x-www-form-urlencoded"}
      ]

      case XeroCLI.HTTP.post(@xero_token_url, body, headers) do
        {:ok, %{status_code: 200, body: response_body}} ->
          case Jason.decode(response_body) do
            {:ok, token_data} ->
              # Clean up the code verifier
              XeroCLI.Config.set_setting("pkce_verifier", nil)

              # Add timestamp for token expiration tracking
              token_data_with_timestamp =
                Map.put(token_data, "obtained_at", System.system_time(:second))

              {:ok, token_data_with_timestamp}

            {:error, reason} ->
              {:error, "Failed to parse token response: #{inspect(reason)}"}
          end

        {:ok, %{status_code: status, body: body}} ->
          {:error, "Token exchange failed with status #{status}: #{body}"}

        {:error, reason} ->
          {:error, "Token exchange request failed: #{inspect(reason)}"}
      end
    end
  end

  def refresh_token(refresh_token, client_id) do
    params = %{
      "grant_type" => "refresh_token",
      "refresh_token" => refresh_token,
      "client_id" => client_id
    }

    body = URI.encode_query(params)

    headers = [
      {"content-type", "application/x-www-form-urlencoded"}
    ]

    case XeroCLI.HTTP.post(@xero_token_url, body, headers) do
      {:ok, %{status_code: 200, body: response_body}} ->
        case Jason.decode(response_body) do
          {:ok, token_data} ->
            token_data_with_timestamp =
              Map.put(token_data, "obtained_at", System.system_time(:second))

            {:ok, token_data_with_timestamp}

          {:error, reason} ->
            {:error, "Failed to parse token response: #{inspect(reason)}"}
        end

      {:ok, %{status_code: status, body: body}} ->
        {:error, "Token refresh failed with status #{status}: #{body}"}

      {:error, reason} ->
        {:error, "Token refresh request failed: #{inspect(reason)}"}
    end
  end

  def get_connections(access_token) do
    headers = [
      {"authorization", "Bearer #{access_token}"},
      {"content-type", "application/json"}
    ]

    case XeroCLI.HTTP.get(@xero_connections_url, headers) do
      {:ok, %{status_code: 200, body: response_body}} ->
        case Jason.decode(response_body) do
          {:ok, connections} -> {:ok, connections}
          {:error, reason} -> {:error, "Failed to parse connections: #{inspect(reason)}"}
        end

      {:ok, %{status_code: status, body: body}} ->
        {:error, "Failed to get connections with status #{status}: #{body}"}

      {:error, reason} ->
        {:error, "Connections request failed: #{inspect(reason)}"}
    end
  end

  def token_expired?(credentials) do
    case Map.get(credentials, "obtained_at") do
      nil ->
        true

      obtained_at ->
        expires_in = Map.get(credentials, "expires_in", 1800)
        current_time = System.system_time(:second)
        # Consider token expired if less than 5 minutes remaining
        current_time >= obtained_at + expires_in - 300
    end
  end

  # Generate a random code verifier for PKCE
  defp generate_code_verifier do
    :crypto.strong_rand_bytes(32)
    |> Base.url_encode64(padding: false)
  end

  # Generate code challenge from verifier using SHA256
  defp generate_code_challenge(verifier) do
    :crypto.hash(:sha256, verifier)
    |> Base.url_encode64(padding: false)
  end

  def start_callback_server do
    {:ok, listen_socket} =
      :gen_tcp.listen(8888, [
        :binary,
        packet: :raw,
        active: false,
        reuseaddr: true
      ])

    IO.puts("\n📡 Waiting for OAuth callback on http://localhost:8888/callback...")
    accept_connection(listen_socket)
  end

  defp accept_connection(listen_socket) do
    {:ok, socket} = :gen_tcp.accept(listen_socket)

    case :gen_tcp.recv(socket, 0, 30_000) do
      {:ok, data} ->
        request = to_string(data)
        code = extract_code_from_request(request)

        if code do
          response = """
          HTTP/1.1 200 OK\r
          Content-Type: text/html\r
          Connection: close\r
          \r
          <html>
          <head><title>Xero CLI - Authentication Successful</title></head>
          <body style="font-family: Arial, sans-serif; text-align: center; padding: 50px;">
            <h1 style="color: #13B5EA;">✓ Authentication Successful!</h1>
            <p>You can close this window and return to your terminal.</p>
          </body>
          </html>
          """

          :gen_tcp.send(socket, response)
          :gen_tcp.close(socket)
          :gen_tcp.close(listen_socket)
          {:ok, code}
        else
          error_response = """
          HTTP/1.1 400 Bad Request\r
          Content-Type: text/html\r
          Connection: close\r
          \r
          <html>
          <head><title>Xero CLI - Authentication Failed</title></head>
          <body style="font-family: Arial, sans-serif; text-align: center; padding: 50px;">
            <h1 style="color: #ff0000;">✗ Authentication Failed</h1>
            <p>No authorization code found. Please try again.</p>
          </body>
          </html>
          """

          :gen_tcp.send(socket, error_response)
          :gen_tcp.close(socket)
          :gen_tcp.close(listen_socket)
          {:error, "No authorization code found"}
        end

      {:error, reason} ->
        :gen_tcp.close(socket)
        :gen_tcp.close(listen_socket)
        {:error, "Failed to receive callback: #{inspect(reason)}"}
    end
  end

  defp extract_code_from_request(request) do
    case Regex.run(~r/GET \/callback\?code=([^&\s]+)/, request) do
      [_, code] -> URI.decode(code)
      _ -> nil
    end
  end
end

File: internal/oauth/oauth.go
package oauth

import (
	"context"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os"
	"time"

	"github.com/msmithstubbs/xero-cli/internal/credentials"
)

const (
	xeroAuthURL        = "https://login.xero.com/identity/connect/authorize"
	xeroTokenURL       = "https://identity.xero.com/connect/token"
	xeroConnectionsURL = "https://api.xero.com/connections"
	redirectURI        = "http://localhost:8888/callback"
)

type TokenData struct {
	AccessToken  string `json:"access_token"`
	RefreshToken string `json:"refresh_token"`
	ExpiresIn    int64  `json:"expires_in"`
	ObtainedAt   int64  `json:"obtained_at"`
}

type Connection struct {
	TenantID   string `json:"tenantId"`
	TenantName string `json:"tenantName"`
}

func GetAuthURL(clientID string) (string, string, error) {
	codeVerifier, err := codeVerifierFromEnv()
	if err != nil {
		return "", "", err
	}
	codeChallenge := generateCodeChallenge(codeVerifier)

	params := url.Values{}
	params.Set("response_type", "code")
	params.Set("client_id", clientID)
	params.Set("redirect_uri", redirectURI)
	params.Set("scope", "offline_access openid profile email accounting.transactions accounting.contacts accounting.settings")
	params.Set("code_challenge", codeChallenge)
	params.Set("code_challenge_method", "S256")

	return fmt.Sprintf("%s?%s", xeroAuthURL, params.Encode()), codeVerifier, nil
}

func ExchangeCode(code, clientID, codeVerifier string) (*TokenData, error) {
	if codeVerifier == "" {
		return nil, errors.New("code verifier not found. Please restart the authentication process")
	}

	form := url.Values{}
	form.Set("grant_type", "authorization_code")
	form.Set("client_id", clientID)
	form.Set("code", code)
	form.Set("redirect_uri", redirectURI)
	form.Set("code_verifier", codeVerifier)

	data, err := postForm(xeroTokenURL, form)
	if err != nil {
		return nil, err
	}
	data.ObtainedAt = time.Now().Unix()
	return data, nil
}

func RefreshToken(refreshToken, clientID string) (*TokenData, error) {
	form := url.Values{}
	form.Set("grant_type", "refresh_token")
	form.Set("refresh_token", refreshToken)
	form.Set("client_id", clientID)

	data, err := postForm(xeroTokenURL, form)
	if err != nil {
		return nil, err
	}
	data.ObtainedAt = time.Now().Unix()
	return data, nil
}

func GetConnections(accessToken string) ([]Connection, error) {
	req, err := http.NewRequest(http.MethodGet, xeroConnectionsURL, nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("authorization", "Bearer "+accessToken)
	req.Header.Set("content-type", "application/json")

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return nil, fmt.Errorf("failed to get connections with status %d: %s", resp.StatusCode, string(body))
	}

	var connections []Connection
	if err := json.Unmarshal(body, &connections); err != nil {
		return nil, err
	}
	return connections, nil
}

func TokenExpired(creds *credentials.Credentials) bool {
	if creds == nil || creds.ObtainedAt == 0 {
		return true
	}
	expiresIn := creds.ExpiresIn
	if expiresIn == 0 {
		expiresIn = 1800
	}
	current := time.Now().Unix()
	return current >= creds.ObtainedAt+expiresIn-300
}

func codeVerifierFromEnv() (string, error) {
	if env := os.Getenv("XERO_PKCE_VERIFIER"); env != "" {
		return env, nil
	}
	return generateCodeVerifier()
}

type CallbackServer struct {
	Server *http.Server
	CodeCh chan string
	ErrCh  chan error
}

func StartCallbackServer(ctx context.Context) (*CallbackServer, error) {
	codeCh := make(chan string, 1)
	errCh := make(chan error, 1)

	mux := http.NewServeMux()
	mux.HandleFunc("/callback", func(w http.ResponseWriter, r *http.Request) {
		code := r.URL.Query().Get("code")
		if code == "" {
			w.WriteHeader(http.StatusBadRequest)
			io.WriteString(w, failureHTML())
			errCh <- errors.New("no authorization code found")
			return
		}
		io.WriteString(w, successHTML())
		codeCh <- code
	})

	srv := &http.Server{
		Addr:              ":8888",
		Handler:           mux,
		ReadHeaderTimeout: 5 * time.Second,
	}

	go func() {
		<-ctx.Done()
		_ = srv.Shutdown(context.Background())
	}()

	go func() {
		if err := srv.ListenAndServe(); err != nil && !errors.Is(err, http.ErrServerClosed) {
			errCh <- err
		}
	}()

	return &CallbackServer{Server: srv, CodeCh: codeCh, ErrCh: errCh}, nil
}

func postForm(url string, form url.Values) (*TokenData, error) {
	resp, err := http.PostForm(url, form)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return nil, fmt.Errorf("token request failed with status %d: %s", resp.StatusCode, string(body))
	}

	var data TokenData
	if err := json.Unmarshal(body, &data); err != nil {
		return nil, err
	}
	return &data, nil
}

func generateCodeVerifier() (string, error) {
	buf := make([]byte, 32)
	if _, err := rand.Read(buf); err != nil {
		return "", err
	}
	return base64.RawURLEncoding.EncodeToString(buf), nil
}

func generateCodeChallenge(verifier string) string {
	hash := sha256.Sum256([]byte(verifier))
	return base64.RawURLEncoding.EncodeToString(hash[:])
}

func successHTML() string {
	return `<!DOCTYPE html>
<html>
<head><title>Xero CLI - Authentication Successful</title></head>
<body style="font-family: Arial, sans-serif; text-align: center; padding: 50px;">
  <h1 style="color: #13B5EA;">Authentication Successful</h1>
  <p>You can close this window and return to your terminal.</p>
</body>
</html>`
}

func failureHTML() string {
	return `<!DOCTYPE html>
<html>
<head><title>Xero CLI - Authentication Failed</title></head>
<body style="font-family: Arial, sans-serif; text-align: center; padding: 50px;">
  <h1 style="color: #ff0000;">Authentication Failed</h1>
  <p>No authorization code found. Please try again.</p>
</body>
</html>`
}
